packages/token-generation-readmodel-checker/src/utils/utils.ts

```diff
@@ -0,0 +1,1242 @@
+                  descriptorAudience: e.descriptorAudience,
+                  descriptorVoucherLifespan: e.descriptorVoucherLifespan,
+                };
+
+              console.log(
+                `token-generation-states entry with PK ${e.PK} is incorrect:
+  ${JSON.stringify(wrongTokenGenStatesEntry)}
+                `
+              );
+              return [...acc, wrongTokenGenStatesEntry];
+            }
+          } else {
+            // TokenGenerationStatesConsumerClient with CLIENTKID PK
+            if (
+              TokenGenerationStatesClientKidPurposePK.safeParse(e.PK).success
+            ) {
+              console.log(
+                `token-generation-states entry has PK ${e.PK}, but should have a CLIENTKID PK`
+              );
+              logger.error(
+                `token-generation-states entry has PK ${e.PK}, but should have a CLIENTKID PK`
+              );
+
+              return [
+                ...acc,
+                {
+                  PK: e.PK,
+                  consumerId: e.consumerId,
+                  clientKind: e.clientKind,
+                  publicKey: e.publicKey,
+                  GSIPK_clientId: e.GSIPK_clientId,
+                  GSIPK_kid: e.GSIPK_kid,
+                  GSIPK_clientId_purposeId: e.GSIPK_clientId_purposeId,
+                  GSIPK_purposeId: e.GSIPK_purposeId,
+                  purposeState: e.purposeState,
+                  purposeVersionId: e.purposeVersionId,
+                  GSIPK_consumerId_eserviceId: e.GSIPK_consumerId_eserviceId,
+                  agreementId: e.agreementId,
+                  agreementState: e.agreementState,
+                  GSIPK_eserviceId_descriptorId:
+                    e.GSIPK_eserviceId_descriptorId,
+                  descriptorState: e.descriptorState,
+                  descriptorAudience: e.descriptorAudience,
+                  descriptorVoucherLifespan: e.descriptorVoucherLifespan,
+                },
+              ];
+            }
+
+            if (
+              getClientIdFromTokenGenStatesPK(e.PK) !== client.id ||
+              e.consumerId !== client.consumerId ||
+              e.GSIPK_clientId !== client.id ||
+              client.keys.every(
+                (k) => !(k.kid === e.GSIPK_kid && k.encodedPem === e.publicKey)
+              ) ||
+              e.GSIPK_kid !== getKidFromTokenGenStatesPK(e.PK)
+            ) {
+              return [
+                ...acc,
+                {
+                  PK: e.PK,
+                  consumerId: e.consumerId,
+                  clientKind: e.clientKind,
+                  publicKey: e.publicKey,
+                  GSIPK_clientId: e.GSIPK_clientId,
+                  GSIPK_kid: e.GSIPK_kid,
+                },
+              ];
+            }
+          }
+          return acc;
+        })
+        .with({ clientKind: clientKindTokenGenStates.api }, (e) => {
+          if (
+            getClientIdFromTokenGenStatesPK(e.PK) !== client.id ||
```

Comment:
just in case, let's check also the PK format

Reply:
I'm not sure the `safeParse` is needed here. `e` is already of type `TokenGenerationStatesApiClient` at this point and the only possible PK is `CLIENTKID#`.

Reply:
makes sense